core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/TransportListTasksAction.java

```diff
@@ -110,4 +150,215 @@ protected void processTasks(ListTasksRequest request, Consumer<Task> operation)
+            this.task = task;
+            this.request = request;
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void onResponse(ListTasksResponse listTasksResponse) {
+            if (listTasksResponse.getTasks().isEmpty()) {
+                startGet();
+            } else {
+                delegate.onResponse(listTasksResponse);
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            if (e instanceof NoSuchNodeException) {
+                startGet();
+            }
+            delegate.onFailure(e);
+        }
+
+        private void startGet() {
+            GetRequest get = new GetRequest(TaskResultsService.TASK_RESULT_INDEX, TaskResultsService.TASK_RESULT_TYPE,
+                    request.getTaskId().toString());
+            get.setParentTask(clusterService.localNode().getId(), task.getId());
+            client.get(get, new ActionListener<GetResponse>() {
+                @Override
+                public void onResponse(GetResponse getResponse) {
+                    try {
+                        if (getResponse.isExists()) {
+                            onNonEmptyGetResponse(getResponse);
+                        } else {
+                            // Didn't find the task either way so just return the empty response
+                            delegate.onResponse(new ListTasksResponse(emptyList(), emptyList(), emptyList()));
+                        }
+                    } catch (Exception e) {
+                        delegate.onFailure(e);
+                    }
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    delegate.onFailure(e);
+                }
+            });
+        }
+
+        private void onNonEmptyGetResponse(GetResponse response) throws IOException {
+            if (response.isSourceEmpty()) {
+                throw new IllegalArgumentException("Stored task status for [" + response.getId() + "] didn't contain any source!");
+            }
+            TaskInfo taskInfo;
+            try (XContentParser parser = XContentHelper.createParser(response.getSourceAsBytesRef())) {
+                taskInfo = TaskInfoBuilder.PARSER.parse(parser, () -> ParseFieldMatcher.STRICT).build();
+                delegate.onResponse(new ListTasksResponse(singletonList(taskInfo), emptyList(), emptyList()));
+            }
+        }
+    }
+
+    /**
+     * Builds TaskInfo from persisted copy in the task result index.
+     */
+    public static class TaskInfoBuilder {
+        public static final ObjectParser<TaskInfoBuilder, ParseFieldMatcherSupplier> PARSER = new ObjectParser<>("task_result",
+                TaskInfoBuilder::new);
+        static {
+            // Note for the future: this has to be backwards compatible with all changes to the task persistence format
```

Comment:
It might be a good idea to store a version of elasticsearch that generated the task somewhere on the top level. So, we could filter out incompatible versions in the future and don't have to support old persistence formats forever.